File: mobile/plugins/sorcha-proximity/android/src/main/java/dev/sorcha/proximity/ProximityCentral.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Sorcha Contributors

package dev.sorcha.proximity

import android.annotation.SuppressLint
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.ParcelUuid
import java.util.UUID
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * The **reader** side of the radio: scans for the holder's advertised service, connects, and moves opaque bytes.
 */
@SuppressLint("MissingPermission") // The plugin probes and requests permissions before ever calling in here.
class ProximityCentral(
    private val context: Context,
    private val onMessage: (ByteArray) -> Unit,
    private val onDisconnected: (String) -> Unit,
    private val onError: (String) -> Unit
) {

    private val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager

    private var gatt: BluetoothGatt? = null
    private var client2Server: BluetoothGattCharacteristic? = null
    private var stateCharacteristic: BluetoothGattCharacteristic? = null

    private val assembler = MessageAssembler()
    private val outbound = ConcurrentLinkedQueue<ByteArray>()

    @Volatile
    private var mtu = 23

    @Volatile
    private var writing = false

    private var serviceUuid: UUID? = null
    private var scanCallback: ScanCallback? = null

    fun connect(serviceUuid: UUID) {
        this.serviceUuid = serviceUuid
        assembler.reset()
        outbound.clear()

        val adapter = manager.adapter
        if (adapter == null || !adapter.isEnabled) {
            onError("Bluetooth is switched off.")
            return
        }

        // Scan only for this session's service. The UUID is random per exchange, so this is both the
        // narrowest possible scan and the one that cannot pick up somebody else's presentation.
        val filter = ScanFilter.Builder()
            .setServiceUuid(ParcelUuid(serviceUuid))
            .build()

        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        val callback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                // First match wins — the service UUID is unique to this exchange, so there is nothing to
                // choose between.
                stopScanning()
                gatt = result.device.connectGatt(context, false, gattCallback)
            }

            override fun onScanFailed(errorCode: Int) {
                onError("Could not scan for the holder (code $errorCode).")
            }
        }
        scanCallback = callback

        adapter.bluetoothLeScanner?.startScan(listOf(filter), settings, callback)
            ?: onError("This device cannot scan for BLE.")
    }

    fun send(message: ByteArray) {
        val characteristic = client2Server
        if (characteristic == null) {
            onError("Not connected to a holder.")
            return
        }

        outbound.addAll(MdocGatt.chunk(message, mtu))
        drain()
    }

    /**
     * Writes one chunk and waits for onCharacteristicWrite before the next.
     *
     * Android allows exactly one outstanding GATT write; queueing them in a loop drops all but the first, and
     * does so intermittently.
     */
    @Synchronized
    private fun drain() {
        if (writing) return

        val characteristic = client2Server ?: return
        val chunk = outbound.poll() ?: return

        writing = true

        @Suppress("DEPRECATION")
        characteristic.value = chunk
        @Suppress("DEPRECATION")
        characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
        @Suppress("DEPRECATION")
        gatt?.writeCharacteristic(characteristic)
    }

    private fun stopScanning() {
        scanCallback?.let { manager.adapter?.bluetoothLeScanner?.stopScan(it) }
        scanCallback = null
    }

    fun stop() {
        stopScanning()
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        client2Server = null
        stateCharacteristic = null
        outbound.clear()
        assembler.reset()
    }

    private val gattCallback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothGatt.STATE_CONNECTED ->
                    // Ask for a bigger MTU before anything else: at the 23-byte default, a DeviceResponse
                    // would be hundreds of packets.
                    gatt.requestMtu(517)

                BluetoothGatt.STATE_DISCONNECTED ->
                    onDisconnected("peerDisconnected")
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, newMtu: Int, status: Int) {
            // The ATT header eats 3 bytes of every packet.
            mtu = newMtu - 3
            gatt.discoverServices()
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            val service = serviceUuid?.let { gatt.getService(it) }
            if (service == null) {
                onDisconnected("transportError")
                return
            }

            client2Server = service.getCharacteristic(MdocGatt.CLIENT_2_SERVER)
            stateCharacteristic = service.getCharacteristic(MdocGatt.STATE)

            val notify = service.getCharacteristic(MdocGatt.SERVER_2_CLIENT)
            if (notify == null || client2Server == null) {
                onDisconnected("transportError")
                return
            }

            // Subscribing is what tells the holder a reader has arrived. Both halves are required: the local
            // enable, AND the CCC descriptor write that actually reaches the peer.
            gatt.setCharacteristicNotification(notify, true)

            val ccc = notify.getDescriptor(MdocGatt.CCC_DESCRIPTOR)
            if (ccc != null) {
                @Suppress("DEPRECATION")
                ccc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                @Suppress("DEPRECATION")
                gatt.writeDescriptor(ccc)
            }
        }

        override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            // Notifications are live. Announce the start explicitly, so the holder is not left inferring it.
            stateCharacteristic?.let {
                @Suppress("DEPRECATION")
                it.value = byteArrayOf(MdocGatt.STATE_START)
                @Suppress("DEPRECATION")
                gatt.writeCharacteristic(it)
            }
        }

        override fun onCharacteristicWrite(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (characteristic.uuid != MdocGatt.CLIENT_2_SERVER) return

            synchronized(this@ProximityCentral) {
                writing = false
            }
            drain()   // the previous chunk is away; send the next
        }

        @Suppress("DEPRECATION")
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            if (characteristic.uuid != MdocGatt.SERVER_2_CLIENT) return
            assembler.accept(characteristic.value)?.let(onMessage)
        }
    }
}

File: mobile/plugins/sorcha-proximity/ios/Sources/SorchaProximity/ProximityCentral.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Sorcha Contributors

import CoreBluetooth
import Foundation

/// The **reader** side of the radio: scans for the holder's advertised service, connects, and moves opaque bytes.
final class ProximityCentral: NSObject {

    private var manager: CBCentralManager?
    private var serviceUuid: CBUUID?

    private var peripheral: CBPeripheral?
    private var client2Server: CBCharacteristic?

    private let assembler = MessageAssembler()

    /// Whole messages received from the holder.
    var onMessage: ((Data) -> Void)?

    /// The channel ended.
    var onDisconnected: ((String) -> Void)?

    /// The manager reached a terminal error state before we could connect.
    var onError: ((String) -> Void)?

    private var pendingStart: (() -> Void)?

    func connect(serviceUuid: CBUUID) {
        self.serviceUuid = serviceUuid
        assembler.reset()

        pendingStart = { [weak self] in
            guard let self, let serviceUuid = self.serviceUuid else { return }
            // Scan only for this session's service. The UUID is random per exchange, so this is both the
            // narrowest possible scan and the one that cannot pick up somebody else's presentation.
            self.manager?.scanForPeripherals(withServices: [serviceUuid])
        }

        manager = CBCentralManager(delegate: self, queue: nil)
    }

    func send(_ message: Data) {
        guard let peripheral, let characteristic = client2Server else {
            onError?("Not connected to a holder.")
            return
        }

        // .withResponse: the ATT layer then flow-controls for us, so a large message cannot outrun the link.
        let mtu = peripheral.maximumWriteValueLength(for: .withResponse)

        for chunk in MdocGatt.chunk(message, mtu: mtu) {
            peripheral.writeValue(chunk, for: characteristic, type: .withResponse)
        }
    }

    func stop() {
        if let peripheral {
            manager?.cancelPeripheralConnection(peripheral)
        }
        manager?.stopScan()
        manager = nil
        peripheral = nil
        client2Server = nil
        assembler.reset()
    }
}

extension ProximityCentral: CBCentralManagerDelegate {

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            pendingStart?()
            pendingStart = nil

        case .unauthorized:
            onError?("Bluetooth permission has not been granted.")

        case .poweredOff:
            onError?("Bluetooth is switched off.")

        case .unsupported:
            onError?("This device has no Bluetooth radio.")

        default:
            break
        }
    }

    func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        // First match wins. The service UUID is unique to this exchange, so there is nothing to choose between.
        central.stopScan()

        self.peripheral = peripheral
        peripheral.delegate = self
        central.connect(peripheral)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        guard let serviceUuid else { return }
        peripheral.discoverServices([serviceUuid])
    }

    func centralManager(
        _ central: CBCentralManager,
        didDisconnectPeripheral peripheral: CBPeripheral,
        error: Error?
    ) {
        onDisconnected?("peerDisconnected")
    }

    func centralManager(
        _ central: CBCentralManager,
        didFailToConnect peripheral: CBPeripheral,
        error: Error?
    ) {
        onDisconnected?("transportError")
    }
}

extension ProximityCentral: CBPeripheralDelegate {

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil, let service = peripheral.services?.first else {
            onDisconnected?("transportError")
            return
        }

        peripheral.discoverCharacteristics(
            [MdocGatt.state, MdocGatt.client2Server, MdocGatt.server2Client],
            for: service)
    }

    func peripheral(
        _ peripheral: CBPeripheral,
        didDiscoverCharacteristicsFor service: CBService,
        error: Error?
    ) {
        guard error == nil, let characteristics = service.characteristics else {
            onDisconnected?("transportError")
            return
        }

        for characteristic in characteristics {
            switch characteristic.uuid {
            case MdocGatt.client2Server:
                client2Server = characteristic

            case MdocGatt.server2Client:
                // Subscribing is what tells the holder a reader has arrived — it is the holder's
                // didSubscribeTo callback, and therefore the real start of the session.
                peripheral.setNotifyValue(true, for: characteristic)

            case MdocGatt.state:
                // Announce the start explicitly, so the holder is not left inferring it.
                peripheral.writeValue(Data([MdocGatt.stateStart]), for: characteristic, type: .withResponse)

            default:
                break
            }
        }
    }

    func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        guard characteristic.uuid == MdocGatt.server2Client,
              let value = characteristic.value else { return }

        if let message = assembler.accept(value) {
            onMessage?(message)
        }
    }
}
